rllib/core/rl_module/rl_module.py

```diff
@@ -0,0 +1,166 @@
+        The default implementation requires the forward_exploration to reutn a dict
+        that has `action_dist` key and its value is an instance of
+        `Distribution`. This assumption must always hold.
+        """
+        return ModelSpec({"action_dist": Distribution})
+
+    def output_specs_train(self) -> ModelSpec:
+        """Returns the output specs of the forward_train method."""
+        return ModelSpec()
+
+    def input_specs_inference(self) -> ModelSpec:
+        """Returns the input specs of the forward_inference method."""
+        return ModelSpec()
+
+    def input_specs_exploration(self) -> ModelSpec:
+        """Returns the input specs of the forward_exploration method."""
+        return ModelSpec()
+
+    def input_specs_train(self) -> ModelSpec:
+        """Returns the input specs of the forward_train method."""
+        return ModelSpec()
+
+    @check_specs(
+        input_spec="input_specs_inference", output_spec="output_specs_inference"
+    )
+    def forward_inference(self, batch: SampleBatchType, **kwargs) -> Mapping[str, Any]:
+        """Forward-pass during evaluation, called from the sampler. This method should
+        not be overriden. Instead, override the _forward_inference method."""
+        return self._forward_inference(batch, **kwargs)
+
+    @abc.abstractmethod
+    def _forward_inference(self, batch: NestedDict, **kwargs) -> Mapping[str, Any]:
+        """Forward-pass during evaluation"""
+
+    @check_specs(
+        input_spec="input_specs_exploration", output_spec="output_specs_exploration"
+    )
+    def forward_exploration(
+        self, batch: SampleBatchType, **kwargs
+    ) -> Mapping[str, Any]:
+        """Forward-pass during exploration, called from the sampler. This method should
+        not be overriden. Instead, override the _forward_exploration method."""
+        return self._forward_exploration(batch, **kwargs)
+
+    @abc.abstractmethod
+    def _forward_exploration(self, batch: NestedDict, **kwargs) -> Mapping[str, Any]:
+        """Forward-pass during exploration"""
+
+    @check_specs(input_spec="input_specs_train", output_spec="output_specs_train")
+    def forward_train(self, batch: SampleBatchType, **kwargs) -> Mapping[str, Any]:
+        """Forward-pass during training called from the trainer. This method should
+        not be overriden. Instead, override the _forward_train method."""
+        return self._forward_train(batch, **kwargs)
+
+    @abc.abstractmethod
+    def _forward_train(self, batch: NestedDict, **kwargs) -> Mapping[str, Any]:
+        """Forward-pass during training"""
+
+    @abc.abstractmethod
+    def get_state(self) -> Mapping[str, Any]:
+        """Returns the state dict of the module."""
+
+    @abc.abstractmethod
+    def set_state(self, state_dict: Mapping[str, Any]) -> None:
+        """Sets the state dict of the module."""
+
+    @abc.abstractmethod
+    def make_distributed(self, dist_config: Mapping[str, Any] = None) -> None:
```

Comment:
We will thoroughly design test and implement this when @avnishn 's trainer PR is made.